File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
    const n = numbers.length;
    const answer = new Array(n).fill(-1); // 기본값 -1
    const stack = []; // 인덱스를 저장할 스택

    for (let i = 0; i < n; i++) {
        // 현재 숫자가 스택 top의 숫자보다 크면 뒷 큰수 발견
        while (stack.length > 0 && numbers[stack[stack.length - 1]] < numbers[i]) {
            const index = stack.pop();
            answer[index] = numbers[i];
        }
        stack.push(i); // 현재 인덱스 push
    }

    return answer;
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Stack<Integer> stack = new Stack<>();

        // 배열의 뒤에서부터 순회
        for (int i = numbers.length - 1; i >= 0; i--) {
            // 스택이 비지 않았고, 현재 수가 스택 최상단 수보다 크면 pop
            while (!stack.isEmpty() && stack.peek() <= numbers[i]) {
                stack.pop();
            }

            // 스택이 비어있으면 뒷큰수가 없음 (-1)
            if (stack.isEmpty()) {
                answer[i] = -1;
            } else {
                // 스택 최상단 수가 뒷큰수
                answer[i] = stack.peek();
            }

            // 현재 수를 스택에 추가
            stack.push(numbers[i]);
        }
        return answer;
    }
}
